test/e2e/upgrade/dns/dns.go

```diff
@@ -0,0 +1,146 @@
+package dns
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/onsi/ginkgo"
+	"github.com/openshift/origin/test/extended/dns"
+	"io"
+	"io/ioutil"
+	kappsv1 "k8s.io/api/apps/v1"
+	kapiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	watchtools "k8s.io/client-go/tools/watch"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/upgrades"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	"time"
+)
+
+type UpgradeTest struct{}
+
+const (
+	podWaitTimeout = 1 * time.Minute
+)
+
+func (t *UpgradeTest) Name() string { return "check-for-dns-availability" }
+func (UpgradeTest) DisplayName() string {
+	return "[sig-network-edge] Verify DNS availability during and after upgrade success"
+}
+
+// Setup creates a DaemonSet to verify DNS availability during and after upgrade
+func (t *UpgradeTest) Setup(f *framework.Framework) {
+	ginkgo.By("Setting up upgrade DNS availability test")
+
+	ginkgo.By("Getting DNS Service Cluster IP")
+	dnsServiceIP := t.getServiceIP(f)
+
+	ginkgo.By("Creating a DaemonSet to verify DNS availability")
+	t.createDNSTestDaemonSet(f, dnsServiceIP)
+}
+
+// Test checks for logs from DNS availability test a minute after upgrade finishes
+// to cover during and after upgrade phase, and verifies the results.
+func (t *UpgradeTest) Test(f *framework.Framework, done <-chan struct{}, _ upgrades.UpgradeType) {
+	// Block until upgrade is done
+	ginkgo.By("Waiting for upgrade to finish before checking for DNS availability")
+	<-done
+
+	ginkgo.By("Sleeping for a minute to give it time for verifying DNS after upgrade")
+	time.Sleep(1 * time.Minute)
+
+	ginkgo.By("Validating DNS results")
+	t.validateDNSResults(f)
+}
+
+// getServiceIP gets Cluster IP from DNS Service
+func (t *UpgradeTest) getServiceIP(f *framework.Framework) string {
+	dnsService, err := f.ClientSet.CoreV1().Services("openshift-dns").Get(context.Background(), "dns-default", metav1.GetOptions{})
+	framework.ExpectNoError(err)
+	return dnsService.Spec.ClusterIP
+}
+
+// createDNSTestDaemonSet creates a DaemonSet to test DNS availability
+func (t *UpgradeTest) createDNSTestDaemonSet(f *framework.Framework, dnsServiceIP string) {
+	// daemon sets should auto replace local references
+	cmd := fmt.Sprintf("while true; do dig +noall @%s google.com && echo success || echo fail; done", dnsServiceIP)
```

Comment:
It might be a good idea to add a `sleep` in here. Maybe `sleep 1` just to space out the lookups a bit. We should also consider whether caching is an important aspect here...for example, if the default caching is 3 seconds we could get a cached record for _n_ lookups before receiving a failure.

I also noticed that `dig` doesn't differentiate between a successful lookup and an `NXDOMAIN`.
```
$ dig +noall asdlkfhsdlfhsdaklhfklsjflklskhflsdflkadflhsdafkljhsdakjhfsdf.com
$ echo $?
0
$ dig +noall google.com
$ echo $?
0
```
The only problem with that is we have no way of programmitcally differentiating between a failed lookup for a domain that we know should be successful (e.g. google.com) and a domain that we know should not be successful. To solve this, it might make sense to just do:
```
while true; do dig +short @%s google.com || echo fail; done
```
Then when looping through the pod logs you can validate that a line is an IP address. If a line is "fail" or not an IP, you know something else went wrong and can fail accordingly. Hopefully that all makes sense!

Reply:
I've applied the suggestion. Thanks!